engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/blueprint/NadelExecutionBlueprintFactory.kt

```diff
@@ -567,3 +599,138 @@ private class SharedTypesAnalysis(
         }
     }
 }
+
+private class TypeOwnership(
+    private val overallSchema: GraphQLSchema,
+    private val services: List<Service>,
+    private val typeRenames: List<NadelTypeRenameInstruction>,
+) {
+    fun getTypesToServicesThatDefineThem(): Map<String, Set<Service>> {
+        return getTypesThatServiceDefines()
+            .entries
+            .asSequence()
+            .flatMap { (service, typesDefined) ->
+                typesDefined
+                    .asSequence()
+                    .map {
+                        it to service
+                    }
+            }
+            .groupBy {
+                it.first
+            }
+            .mapValues { it: Map.Entry<String, List<Pair<String, Service>>> ->
+                it.value // Get the List<Pair<…>> out
+                    .asSequence()
+                    .map {
+                        it.second // Get the Service out of the Pair<String, Service>
+                    }
+                    .toSet()
+            }
+    }
+
+    fun getTypesThatServiceDefines(): Map<Service, Set<String>> {
+        return services
+            .map { service ->
+                service to getTypeThatServiceDefined(service)
+            }
+            .toMapStrictly()
+    }
+
+    private fun getTypeThatServiceDefined(service: Service): Set<String> {
+        val typesDefined = service.definitionRegistry.definitions
+            .asSequence()
+            .filterIsInstance<AnyNamedNode>()
+            .filterNot { it.isExtensionDef }
+            .map { it.name }
+            .toHashSet()
+
+        val rawTypesUsedAsOutputTypes = service.definitionRegistry.definitions
+            .asSequence()
+            .filterIsInstance<AnyNamedNode>()
+            .filter {
+                // If the definition is an extension, only consider it if it is extending a type the service owns
+                it.isNotExtensionDef || it.name in typesDefined
+            }
+            .filterIsInstance<AnyImplementingTypeDefinition>()
+            .flatMap {
+                it.fieldDefinitions
+            }
+            .filterNot {
+                isHydrationField(it)
```

Comment:
Output types of hydrated fields must NOT be considered. A service can reference ANY type for a hydration and never define it in their service.